### ISSUE TITLE
Fix model.fit() multiple output

### ISSUE THREAD

**Issue by kirillbondarev** [fchollet/keras#212]
I have used Keras for not so long, but here's the thing.
If you create several models in one script, then you are starting to get multiple output during training, like this (5th model in one session):
Epoch 4
Epoch 4
Epoch 4
Epoch 4
Epoch 4
1664/7000 [======>.......................] - ETA: 116s - loss: 0.4524

After the epoch has ended you alse get several full progressbars. I tried to find the source of problem, but I'm not so acquainted with Keras code. Looks like the problem is somewhere in the callbacks mechanism, but I can't be sure.

**Comment by tkipf** [comment]
Seems like the BaseLogger object is created and appended to CallbackList for each model that is created with "verbose = 1", but is not removed from CallbackList or destroyed after training is finished. This issue also persists after running "%reset" in an interactive ipython session.

**Comment by fchollet** [comment]
Do you have any code to repro?

@tristandeleu what's your take on this?

**Comment by tkipf** [comment]
Here's an example:
```python
from keras.models import Sequential
from keras.layers.core import Dense
from keras.optimizers import SGD
import numpy as np

X = np.zeros((1,1))
model = Sequential()
model.add(Dense(1, 1))
model.compile(loss='mse', optimizer='sgd')
model.fit(X, X, nb_epoch=1)

model2 = Sequential()
model2.add(Dense(1, 1))
model2.compile(loss='mse', optimizer='sgd')
model2.fit(X, X, nb_epoch=1)
```
Produces the following output:
```
Epoch 0
1/1 [==============================] - 0s - loss: 0.0000
Epoch 0
Epoch 0
1/1 [==============================] - 0s - loss: 0.0000
1/1 [==============================] - 0s - loss: 0.0000
```

**Comment by fchollet** [comment]
I'm starting to think that Python is a horrible language. The problem was pointer voodoo. Here's a quick explanation.

The fit method takes a `callbacks` argument, which should be a list of `Callback` objects. The default value is `callbacks=[]`.

This list is used to initialize a `CallbackList` object, which is constructed as:
```python
def __init__(self, callbacks=[], queue_length=10):
    self.callbacks = callbacks
```

At this point the `self.callbacks` variable in our `CallbackList` (scope: `model.fit` call) and the `callbacks` variable (scope: `model.fit` call) are the same pointer. But here's where it gets funny: the keyword argument `callbacks` in any future `models.Sequential` object is also that same pointer! 

So if you pass an empty `callbacks` argument in the fit call to the second model (or any `callbacks` list really):
```python
model2.fit(X, X, nb_epoch=1, callbacks=[])
```
Then logging works as expected. 

If you *don't* pass a `callbacks` argument (the default value is `callbacks=[]`), logging becomes redundant, because the default value for the keyword argument is not being used when no argument is passed, instead what's being used is the `self.callbacks` list that is an attribute of a `CallbackList` that lived in the scope of the `fit` call in a previous `models.Sequential` instance. 

Pure insanity.

The fix: replace 
```python
def __init__(self, callbacks=[], queue_length=10):
    self.callbacks = callbacks
```
with 
```python
def __init__(self, callbacks=[], queue_length=10):
    self.callbacks = [c for c in callbacks] # copy values instead of list pointer
```

**Comment by tristandeleu** [comment]
Very good catch!
That is a side effect I have not really thought about. Sorry about that, I'm glad you quickly found it out.

**Comment by danielhers** [comment]
Hi @fchollet and @tristandeleu,
To copy a list, you can just do `self.callbacks = list(callbacks)`. No need for the list comprehension here.
However, it would be better to use `None` instead. See here: http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments
